dashboard/dashboard.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package dashboard
+
+//go:generate go-bindata -nometadata -o assets.go -prefix assets -pkg dashboard assets/public/...
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rcrowley/go-metrics"
+	"golang.org/x/net/websocket"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	processorSampleLimit = 200
+	memorySampleLimit    = 200
+	trafficSampleLimit   = 200
+)
+
+var (
+	nextId uint32 // Next connection id
+)
+
+type dashboard struct {
+	config *Config
+
+	listener net.Listener
+	conns    map[uint32]*client // Currently live websocket connections
+	Metrics  *metricSamples     `json:"metrics,omitempty"`
+	Stats    *status            `json:"stats,omitempty"`
+	lock     sync.RWMutex       // Lock protecting the dashboard's internals
+
+	closing chan chan error // Channel used for graceful exit
+	wg      sync.WaitGroup
+}
+
+type client struct {
+	conn   *websocket.Conn              // Particular live websocket connection
+	msg    chan *map[string]interface{} // Message queue for the update messages
```

Comment:
You don't need a pointer, map is already a lightweight wrapper.

Reply:
Also please convert the map[string]interface{} to a proper type with typed fields and json tags.